update

```diff
@@ -33,13 +34,22 @@ dirs.forEach(function (dir) {
       return fs.readFileSync(path.join(__dirname, dir, file))
     }
 
+    function readBuiltinVersion (file) {
+      // NOTE: should be using this, but it leaks memory
+      // return solc(require(path.join(__dirname, dir, file))).version()
+      const filename = path.join(__dirname, dir, file)
+      return cp.execSync(`/usr/bin/env node -e "var solc = require('${filename}'); console.log(solc.cwrap(('_solidity_version' in solc) ? 'solidity_version' : 'version', 'string', [])())"`).toString().trim()
```

Comment:
It looks like it's going to be very slow. When I was rewriting bytecode comparison scripts and I tried doing something like this, rerunning `node` for each compilation was very slow (https://github.com/ethereum/solidity/issues/10165#issuecomment-744521194). A loop in JS was one or two orders of magnitude faster.

Is it actually a leak or is it just that `update` itself is pretty bad about controlling its memory usage (https://github.com/ethereum/solidity/issues/9956)? Because for the bytecode comparison I run [prepare_report.js](https://github.com/ethereum/solidity/blob/v0.8.2/scripts/bytecodecompare/prepare_report.js) which compiles around 5k test cases in a loop, twice and it does not run out of memory in the [t-bytecode-compare.yml](https://github.com/ethereum/solc-bin/blob/gh-pages/.github/workflows/t-bytecode-compare.yml) action.

Reply:
Nodejs is bad at controlling memory. We set the loaded emscripten modules to empty reference, but it still won't be GC'd in time somehow. Perhaps it is not a problem anymore with the wasm binaries?

Reply:
ok, right. It's about loading the compiler into memory, not running it. Bytecode comparison loads only one version.

> Perhaps it is not a problem anymore with the wasm binaries?

We still need to run it on all those old asm.js nightlies so I think this does not help us even if it's fixed for wasm. Unless we just compute versions for asm.js once and cache them.

The script now has an option that makes it not recalculate hashes. It could have a similar one for this. The downside would be that if you replace the binary, CI won't update the versions so it would be safer to somehow combine it with detecting which binaries changed in the PR (which adds a bit of complexity to the whole setup).